File: ts/test/c2020_12/c2020_12_23.ts
import {expect} from 'chai';

/**
 * ### Next Greater Element III
 *
 * {@link https://leetcode.com/explore/challenge/card/december-leetcoding-challenge/572/week-4-december-22nd-december-28th/3578/}
 */
function nextGreaterElement(n: number): number {
    const MAX = 2_147_483_647; // 2^31 - 1
    if (n < 12) {
        return -1;
    }

    const chars = n.toString().split('');

    for (let r = chars.length - 1; r > 0; r--) {
        const l = r - 1;
        if (chars[l] < chars[r]) {
            let swap = chars.length - 1;
            while (chars[l] >= chars[swap]) swap--;
            [chars[l], chars[swap]] = [chars[swap], chars[l]];

            const head = chars.slice(0, r).join('');
            const tail = chars.slice(r).sort().join('');

            const result = parseInt(head + tail, 10);

            return result > MAX ? -1 : result;
        }
    }

    return -1;
}

describe('c2020_12_23 Next Greater Element III', () => {
    it('Example 1: (n = 12) -> 21', () => {
        expect(nextGreaterElement(12)).to.eql(21);
    });
    it('Example 2: (n = 21) -> -1', () => {
        expect(nextGreaterElement(21)).to.eql(-1);
    });

    it('Test 36: (n = 1999999999) -> -1', () => {
        expect(nextGreaterElement(1999999999)).to.eql(-1);
    });

    it('(n = 14654321) -> 15123446', () => {
        expect(nextGreaterElement(14654321)).to.eql(15123446);
    });
    it('(n = 2147483647) -> -1', () => {
        expect(nextGreaterElement(2_147_483_647)).to.eql(-1);
    });
});

File: rust/src/c2020_12/c2020_12_23.rs
#![allow(dead_code)]

/// ### Next Greater Element III
///
/// https://leetcode.com/explore/challenge/card/december-leetcoding-challenge/572/week-4-december-22nd-december-28th/3578/
struct Solution;

//noinspection DuplicatedCode
impl Solution {
    pub fn next_greater_element(n: i32) -> i32 {
        if n < 12 {
            -1
        } else {
            let mut bytes = n.to_string().bytes().collect::<Vec<_>>();

            for r in (1..bytes.len()).rev() {
                let l = r - 1;
                if bytes[l] < bytes[r] {
                    let s = bytes.iter().rposition(|&b| bytes[l] < b).unwrap();
                    bytes.swap(l, s);

                    (&mut bytes[r..]).sort();

                    return String::from_utf8(bytes).expect("Illegal bytes")
                        .parse().unwrap_or(-1);
                }
            }

            -1
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn example1_n12_is_21() { assert_eq!(Solution::next_greater_element(12), 21); }

    #[test]
    fn example2_n21_is_m1() { assert_eq!(Solution::next_greater_element(21), -1); }

    #[test]
    fn n21321_is_22113() { assert_eq!(Solution::next_greater_element(21321), 22113); }

    #[test]
    fn n2147483647_is_m1() { assert_eq!(Solution::next_greater_element(2147483647), -1); }
}
